### PR TITLE
Remove unnecessary GetTypeInfo from Microsoft.Extensions.

### PR DESCRIPTION
**Pull request by eerhardt**
I also made a slight optimization to CallSiteFactory to use ToArray instead of ToList.

These were showing up as 5% of `ServiceCollection.BuildServiceProvider()` in a simple benchmark:

image

```C#
    public class DIBench
    {
        private readonly ServiceCollection services = new ServiceCollection();

        public DIBench()
        {
            services.AddTransient<Service1>();
            services.AddTransient<Service2a>();
            services.AddTransient<Service2b>();
            services.AddTransient<Service3a>();
            services.AddTransient<Service4>();
            services.AddTransient<Service4>();
            services.AddTransient<Service5>();
            services.AddTransient<Service6>();
            services.AddTransient<Service7>();
            services.AddTransient<Service8>();
            services.AddTransient<Service9>();
            services.AddTransient<Service10>();
            services.AddTransient<Service11>();
            services.AddTransient<Service12>();
            services.AddTransient<Service13>();
            services.AddTransient<Service14>();
            services.AddTransient<Service15>();
            services.AddTransient<Service16>();
            services.AddTransient<Service17>();
            services.AddTransient<Service18>();
            services.AddTransient<Service19>();
            services.AddTransient<Service20>();
            services.AddTransient<Service21>();
            services.AddTransient<Service22>();
            services.AddTransient<Service23>();
            services.AddTransient<Service24>();
            services.AddTransient<Service25>();
            services.AddTransient<Service26>();
            services.AddTransient<Service27>();
            services.AddTransient<Service28>();
            services.AddTransient<Service29>();
        }

        [Benchmark]
        public object BuildServiceProvider()
        {
            var sp = services.BuildServiceProvider();

            return sp;
        }
    }
```

### CODE REVIEW

**Comment by davidfowl** [comment]
Nice clean up. We should do this in ASP.NET too...

**Comment by eerhardt** [comment]
> We should do this in ASP.NET too...

Will log an issue (and maybe make a PR if it looks quick).

**Review by stephentoub — APPROVED** [review]
(Some of this LINQ usage pains me to look at :), but that's not related to your PR.)

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs [hunk 1]
```diff
@@ -24,27 +24,27 @@ internal class CallSiteFactory
         public CallSiteFactory(IEnumerable<ServiceDescriptor> descriptors)
         {
             _stackGuard = new StackGuard();
-            _descriptors = descriptors.ToList();
+            _descriptors = descriptors.ToArray();
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:27]: The benefit was very minimal, probably not even observable. I started this change by looking at `CallSiteFactory.Populate` (which is 75% of the above benchmark). And this `ToList()` jumped out at me and looking at the code I noticed we don't need a List, but can just use an Array instead. I just figured while I was in here, it was one less object to create.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:27]: >  I just figured while I was in here, it was one less object to create.

But unless the array is the exact right size,  ToArray needs to allocate one at the end and copy all the elements to it.  ToList doesn't. So you still generally end up with another object,  but typically a bigger one.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:27]: In this case, the Type that is passed in to `descriptors` is always an [`IServiceCollection`](https://github.com/dotnet/runtime/blob/88ecbab85beead9c7d9406ee2f3ea468b1d4cd1e/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/IServiceCollection.cs#L11), which implements `ICollection<T>`. So the size will be known up-front and only a single Array needs to be allocated. It won't need to be resized in the end.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs:27]: Cool

**Pull request merged**